devito/petsc/solve.py

```diff
@@ -83,38 +85,77 @@ def separate_eqn(eqn, target):
     where F(target) = b.
     """
     zeroed_eqn = Eq(eqn.lhs - eqn.rhs, 0)
-    tmp = eval_time_derivatives(zeroed_eqn.lhs)
-    b, F_target = remove_target(tmp, target)
-    return -b, F_target
+    zeroed_eqn = eval_time_derivatives(zeroed_eqn.lhs)
+    target_funcs = set(generate_targets(zeroed_eqn, target))
+    b, F_target = remove_target(zeroed_eqn, target_funcs)
+    return -b, F_target, target_funcs
+
+
+def generate_targets(eq, target):
+    """
+    Extract all the functions that share the same time index as the target,
+    but may have different spatial indices.
+    """
+    funcs = retrieve_functions(eq)
+    if any(dim.is_Time for dim in target.dimensions):
+        time_idx = [
+            i for i, d in zip(target.indices, target.dimensions) if d.is_Time
+        ]
+        targets = [
+            func for func in funcs
+            if func.function is target.function and time_idx[0]
+            in func.indices
+        ]
+    else:
+        targets = [
+            func for func in funcs
```

Comment:
if you use a single letter for this kind of simple comprehensions, like `f`, they will fit in a single line I think

same story with dim, u may just use `d`

Reply:
but the most important thing is consistency across your code... and even better, consistency across the entire codebase

hence single letters generally preferable 

note this is a nitpick